packages/salesforce-adapter/src/change_validators/flow_referenced_elements.ts

```diff
@@ -0,0 +1,122 @@
+  const targetReferences = new Map<string, ElemID>()
+  const findFlowNodesAndTargetReferences: TransformFuncSync = ({ value, path, field }) => {
+    if (!field || !path) return value
+    if (isFlowNodeName(value, path, field.parent)) {
+      flowNodes.set(value, path)
+    }
+    if (isTargetReference(value, path)) {
+      targetReferences.set(value, path)
+    }
+    return value
+  }
+  transformValuesSync({
+    values: element.value,
+    pathID: element.elemID,
+    type: element.getTypeSync(),
+    transformFunc: findFlowNodesAndTargetReferences,
+  })
+  return { flowNodes, targetReferences }
+}
+
+const hasMissingReferencedElements = (targetReferences: string[], flowNodes: string[]): boolean =>
+  !targetReferences.every(targetReference => targetReference in flowNodes)
+
+const hasUnreferencedElements = (targetReferences: string[], flowNodes: string[]): boolean =>
+  !flowNodes.every(flowNode => flowNode in targetReferences)
+
+const hasFlowReferenceError = ({
+  targetReferences,
+  flowNodes,
+}: {
+  targetReferences: Map<string, ElemID>
+  flowNodes: Map<string, ElemID>
+}): boolean =>
+  hasMissingReferencedElements(Array.from(targetReferences.keys()), Array.from(flowNodes.keys())) ||
+  hasUnreferencedElements(Array.from(targetReferences.keys()), Array.from(flowNodes.keys()))
+
+const createMissingReferencedElementChangeError = (targetReference: string, elemId: ElemID): ChangeError => ({
+  elemID: elemId,
+  severity: 'Error',
+  message: 'Reference to missing Flow Element',
+  detailedMessage: `The Flow Element "${targetReference}" does not exist.`,
+})
+
+const createUnreferencedElementChangeError = (flowNode: string, elemId: ElemID): ChangeError => ({
+  elemID: elemId,
+  severity: 'Info',
+  message: 'Unused Flow Element',
+  detailedMessage: `The Flow Element “${flowNode}” isn’t being used in the Flow.`,
+})
+
+const createChangeError = ({
+  targetReferences,
+  flowNodes,
+}: {
+  targetReferences: Map<string, ElemID>
+  flowNodes: Map<string, ElemID>
+}): ChangeError[] => {
+  const errors: ChangeError[] = []
+  const isReferenceToMissingElement = ([reference, _elemId]: [string, ElemID]): boolean => !flowNodes.has(reference)
+  const missingReferencedElementsErrors = Array.from(targetReferences).filter(isReferenceToMissingElement)
+  const isUnreferencedElement = ([reference, _elemId]: [string, ElemID]): boolean => !targetReferences.has(reference)
+  const unreferencedElementsErrors = Array.from(flowNodes).filter(isUnreferencedElement)
+  if (missingReferencedElementsErrors.length > 0) {
+    missingReferencedElementsErrors.forEach(([targetReference, elemId]) =>
+      errors.push(createMissingReferencedElementChangeError(targetReference, elemId)),
+    )
+  }
+  if (unreferencedElementsErrors.length > 0) {
+    unreferencedElementsErrors.forEach(([flowNode, elemId]) =>
+      errors.push(createUnreferencedElementChangeError(flowNode, elemId)),
+    )
+  }
+  return errors
```

Comment:
I find this a bit clearer
```suggestion
  const targetEntries = Array.from(targetReferences)
  const flowEntries = Array.from(flowNodes)

  const missingReferenceErrors = targetEntries
    .filter(([reference]) => !flowNodes.has(reference))
    .map(([reference, elemId]) => createMissingReferencedElementChangeError(reference, elemId))

  const unreferencedErrors = flowEntries
    .filter(([reference]) => !targetReferences.has(reference))
    .map(([reference, elemId]) => createUnreferencedElementChangeError(reference, elemId))

  return [...missingReferenceErrors, ...unreferencedErrors]
```
(But if the lists can be very long sometimes `...` is not a good approach and its better to push)